File: 4.depth-first-search/지연/1.increasing-order-search-tree.py
# https://leetcode.com/problems/increasing-order-search-tree/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        tree = TreeNode()
        node = tree

        def inorder(root):
            nonlocal node
            if root:
                inorder(root.left)
                node.right = TreeNode(root.val)
                node = node.right
                inorder(root.right)
        inorder(root)
        return tree.right

File: 4.depth-first-search/동규/1.increasing-order-search-tree.js
// https://leetcode.com/problems/increasing-order-search-tree/
function increasingBST(root) {
  let ans = new TreeNode(0);
  let cur = ans;
  function inorder(node) {
    if (node === null) return;
    inorder(node.left);
    node.left = null;
    cur.right = node;
    cur = node;
    inorder(node.right);
  }
  inorder(root);
  return ans.right;
}
